### DISCUSSION TITLE
How to avoid circular dependencies

### DISCUSSION THREAD

**Discussion by 3ynm** [jeremyevans/sequel#2064] [General]
I tend to use require to make Sequel aware of classes when moving from a parent to a child and vice versa.

But the console output yells at me "warning: loading in progress, circular require considered harmful".

Am I doing it wrong?

**Comment by jeremyevans** [comment]
If you have a circular require situation, then yes, you are doing it wrong.  This isn't a Sequel issue, this is a general Ruby issue.  You can guard requires `require 'library' unless defined?(Something)` to avoid the circular require issue.

**Reply by 3ynm** [reply]
never thought of that, thanks!

**Comment by 3ynm** [comment]
Well, turns out that the circular dependency issue doesn't get solved with the proposed solution, since the class is still not fully loaded (and thus not defined) at the point where it conflicts.

I will simply load all classes beforehand.

I'll leave this open a couple days, just in case.

**Reply by jeremyevans** [reply]
It should always be possible to fix the circular require by guarding the require and potentially moving it to the necessary point in the files.  See https://github.com/rack/rack/commit/cd4b3c934a8681f5d15053295b7dc34d56b2ccad#diff-a021933e2d0d94eb27857ea78ce733a1b867bd326f119f707886e02aa0fa5e38 for an example of a case in Rack where we had to do that.

Loading all classes beforehand is also fine, and probably better if you need all classes loaded anyway.

**Reply by 3ynm** [reply]
Indeed moving the requires within the class definition solve the issue.

It would have taken me much longer to reach this conclusion by myself. Appreciate it a lot.
